### PR TITLE
Http2 v4

### PR DESCRIPTION
**Pull request by catenacyber**
Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:
https://redmine.openinfosecfoundation.org/issues/1947

DRAFT for discussion
Not to be merged

Describe changes:
- Adds basic HTTP2 parser

Much more work to be done :
 + Commit squashing
 + handle transactions with progress completion
 + parse interesting frame types
 + connection upgrade from HTTP1
 + rules file for applayer events
 + tests
 + robustness against evasions

Modifies #4772 with 
- fixing a bit unit tests

### CODE REVIEW

File: src/app-layer-http2.c [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+#include "app-layer-http2.h"
+#include "rust.h"
+
+static int HTTP2RegisterPatternsForProtocolDetection(void)
+{
+    //TODO is this too restrictive and can be evaded ?
+    /* TODO investigate why this makes AppLayerTest09 fail/leak
+     * and not leak if we only give a 5-bytes pattern
+    if (AppLayerProtoDetectPMRegisterPatternCI(IPPROTO_TCP, ALPROTO_HTTP2,
```
Comment (catenacyber) [inline src/app-layer-http2.c:43]: Any thoughts on what is going on here ?

Comment (catenacyber) [reply src/app-layer-http2.c:43]: (AppLayerTest10 and AppLayerTest11 fail too)
Commenting out `    FAIL_IF(!FLOW_IS_PM_DONE(&f, STREAM_TOSERVER));` makes them succeed again

File: src/detect-http2.c [hunk 2]
```diff
@@ -0,0 +1,612 @@
...
+    if (space) {
+        // a space separates identifier and value
+
+        // copy and isolate first part of string
+        char str_first[HTTP2_MAX_SETTINGS_ID_LEN];
+        if (HTTP2_MAX_SETTINGS_ID_LEN <= space - str) {
+            SCFree(http2set);
+            return -1;
+        }
+        strlcpy(str_first, str, space - str + 1);
```
Comment (victorjulien) [inline src/detect-http2.c:559]: the 3rd arg to strlcpy is the size of the destination buffer, so `sizeof(str_first)`

Comment (catenacyber) [reply src/detect-http2.c:559]: Hmm. I do not want to copy the full buffer, hence the upper check.
`strncpy` was refused by cocci. So I found strlcpy...

And what about Rust for parsing signatures ?

Comment (victorjulien) [reply src/detect-http2.c:559]: No objection to using Rust.

File: src/app-layer-http2.c (hunk 1 above)
Comment (victorjulien) [reply src/app-layer-http2.c:43]: By having a long pattern you force the protocol detection engine to post-pone its judgement about incoming data until its been able this pattern as well.

Comment (catenacyber) [reply src/app-layer-http2.c:43]: Indeed.
But is it good or bad ?

Should not the test end protocol detection by closing the stream anyways ?

File: src/detect-http2.c (hunk 2 above)
Comment (catenacyber) [reply src/detect-http2.c:559]: Some strategy to advise ? Code strangling ?

**Pull request closed without merging**